src/events/event_storage.cpp

```diff
@@ -0,0 +1,75 @@
+//
+// Created by doralitze on 11/15/23.
+//
+
+#include "event_storage.hpp"
+
+#include <limits>
+#include <stdexcept>
+#include "event_source.hpp"
+
+namespace dmxfish::events {
+
+    event_storage::event_storage() : storage_a(), storage_b(), storage_swap_mutex(), ongoing_events() {
+        // Experiments have shown that we grow to a size of approx. 50 very quick and usually stay there
+        storage_a.reserve(50);
+        storage_b.reserve(50);
+    }
+
+    event_storage::~event_storage() {
+        for(auto& registered_source : this->senders) {
+            registered_source->deregister();
+        }
+    }
+
+    [[nodiscard]] bool event_storage::insert_event(const event& e) {
+        if(!this->storage_swap_mutex.try_lock()) {
+            return false;
+        }
+        auto& storage = !this->current_read_storage_is_a ? this->storage_a : this->storage_b;
+        storage.emplace_back(e);
+        this->storage_swap_mutex.unlock();
+        return true;
+    }
+
+    void event_storage::swap_buffers() {
+        auto& old_read_storage = this->current_read_storage_is_a ? this->storage_a : this->storage_b;
+        old_read_storage.clear();
+        this->storage_swap_mutex.lock();
+        this->current_read_storage_is_a = !this->current_read_storage_is_a;
+        this->storage_swap_mutex.unlock();
+
+        auto& new_read_storage = this->current_read_storage_is_a ? this->storage_a : this->storage_b;
+        for(auto& ev : new_read_storage) {
+            switch (ev.get_type()) {
+                case event_type::SINGLE_TRIGGER:
+                case event_type::ONGOING_EVENT:
+                case event_type::INVALID:
+                default:
+                    break;
+                case event_type::START:
+                    this->ongoing_events[ev.get_event_sender()] = event{event_type::ONGOING_EVENT, ev};
+                    break;
+                case event_type::RELEASE:
+                    this->ongoing_events.erase(ev.get_event_sender());
+                    break;
+            }
+        }
+
+        for (auto& lasting_ev : this->ongoing_events) {
+            new_read_storage.push_back(lasting_ev.second);
+        }
```

Comment:
for me it seems a little bit weird, that a event (first time START) is twice in the storage, once with START and once with ONGOING_EVENT.
Is it right that an event started and released in one storage does not get recognized as ongoing_event?
I think it is good, but could do strange things in the filter (if it should get recognized, but one is only looking for the ongoing events)

Reply:
Well, I kinda did that on purpose, as filter might either look for a rising edge or for the continuous presence but usually not for both. In the rare case that they do, they need to they need to implement proper precautions themself.

I also did not implement a lookup index for events as our event objects are small and presumably few. As a consequence I expect linear iteration to be faster than building a hash map every iteration and then performing lookups on them but we need to profile this in the real world.